!!! note  
    Utility scripts are hosted in [the MPH repository](https://github.com/jiang18/mph/tree/main/util).

## GRM input/output
MPH computes only **genomic** relationship matrices. **External general** relationship matrices can be converted to MPH format for use in MPH. 

- `mph_functs.R` contains two R functions for GRM input/output.
    - `write_grm()` writes an R GRM into MPH format files (.iid and .bin).
    - `read_grm()` reads an MPH GRM into R.
    - The R matrix to be written can be any **general** relationship matrix.
    - This script is sourced in other R scripts for GRM input/output.
- `grm_txt2bin.R` converts a relationship matrix from txt to MPH binary format.
    - `Rscript --no-save grm_txt2bin.R example.grm.txt example`
    - This input file should be structured like [**example.grm.txt**](https://github.com/jiang18/mph/blob/main/examples/example.grm.txt).
- `AGHmatrix2mph.R` constructs a **numerator** relationship matrix with [AGHmatrix](https://cran.r-project.org/web/packages/AGHmatrix/) and writes it into MPH format files.
- `make_gci_grm.R` generates a genotypeâ€“covariate interaction GRM from a routine GRM and an indicator matrix.

## Making multivariate files
Four R scripts are provided to prepare input files for multivariate REML in MPH.

- `make_mv_pheno.R` generates multivariate phenotype and covariate files.
    - A covariate file should always be used in MPH's multivariate REML.
    - If **no** covariate file exists initially, use the one generated by `make_mv_pheno.R`.
- `make_mv_covariates.R` generates a multivariate covariate file.
    - If **an** original covariate file exists, use the new one generated by `make_mv_covariates.R`.
- `make_mv_grms.R` generates multivariate GRMs, each corresponding to a specific trait pair.
    - **.G11**: the GRM for the genetic variance of trait 1.
    - **.G12**: the GRM for the genetic covariance of traits 1 and 2.
- `make_mv_erms.R` generates multivariate environmental relationship matrices (ERMs).
    - **.E1d**: the ERM for the difference between the environmental variances of trait 1 and the last trait.
    - **.E12**: the ERM for the environmental covariance of traits 1 and 2.

Usage examples are available [here](examples.md#genetic-correlation).

## Heritability enrichment
### Making a SNP info file
A Perl script, `make_snp_info.pl`, is provided to create [the SNP info file](options.md#snp-info-file).

Usage:

- `perl make_snp_info.pl PLINK-bim-file functional-annotation-file output-filename-prefix`
- Both input files should be in plain text format without header lines.
- Functional-annotation-file's first four columns should be chrom, start, end, and category.

Example:

- `perl make_snp_info.pl test.bim test.annot.txt test`
- The input files are [**test.bim**](https://github.com/jiang18/mph/blob/main/examples/test.bim) and [**test.annot.txt**](https://github.com/jiang18/mph/blob/main/examples/test.annot.txt).
- An output file named **test.snp_info.csv** will be generated. 

### From VCs to enrichments
The **.mq.vc.csv** file produced by `mph --minque` contains the estimates and SEs of VCs, PVEs, and enrichments, where PVEs and enrichments are valid only when functional annotation categories do not overlap with one another.

If functional categories actually overlap, one more quick computation is needed to recompute PVEs and enrichments from **.mq.vc.csv** and [the SNP info file](options.md#snp-info-file). This can be quickly done by `recompute_enrichments()` in [**mph_functs.R**](https://github.com/jiang18/mph/tree/main/util/mph_functs.R).

`recompute_enrichments(vcfile, crossprod, nsnps=NA, annot.size=NA)`

- vcfile: the .mq.vc.csv file produced by `mph --minque`.
- crossprod: the crossproduct of SNP incidence matrix and SNP weighting matrix. Its row names and columns should match the annotation categories of interest and the rows of vcfile, respectively.
- nsnps: the total number of SNPs. If `NA`, it is set to the first GRM's number of SNPs in vcfile.
- annot.size: a list of the number of SNPs in each annotation category listed in the row names of crossprod. It can be computed from the column-wise sum of the corresponding SNP incidence matrix. If `NA`, it is set to the `m` column of vcfile.

A usage example is available [here](examples.md#by-functional-annotations).
