!!! note  
    Utility scripts are hosted in [the MPH repository](https://github.com/jiang18/mph/tree/main/util).

## Utility scripts

### GRM I/O
MPH computes only **genomic** relationship matrices. `External` **general** relationship matrices can be converted to MPH format for use in MPH. 

- `grm_io.R` contains two R functions.
    - `write_grm()` writes an R GRM into MPH format files (.iid and .bin).
    - `read_grm()` reads an MPH GRM into R.
    - The R matrix to be written can be any **general** relationship matrix.
    - This script is sourced in other R scripts for GRM I/O.
- `grm_txt2bin.R` converts a relationship matrix from txt to MPH binary format.
    - `Rscript --no-save grm_txt2bin.R example.grm.txt example`
    - This input file should be structured like [example.grm.txt](https://github.com/jiang18/mph/blob/main/examples/example.grm.txt).
- `AGHmatrix2mph.R` constructs a **numerator** relationship matrix with [AGHmatrix](https://cran.r-project.org/web/packages/AGHmatrix/index.html) and writes it into MPH format files.

### Making files for trait pair
Four R scripts are provided to prepare input files for bivariate REML in MPH.

- `make_pheno_for_pair.R` generates new phenotype and covariate files for a trait pair.
    - A covariate file should always be used in MPH's bivariate REML.
    - If **no** covariate file exists initially, use the one generated by `make_pheno_for_pair.R`.
- `make_covariate_for_pair.R` generates a new covariate file for a trait pair.
    - If **an** original covariate file exists, use the new one generated by `make_covariate_for_pair.R`.
- `make_grms_for_pair.R` generates three GRMs for a trait pair: 
    - One for the variance of trait 1 (**.G11**)
    - One for the variance of trait 2 (**.G22**)
    - One for the covariance between them (**.G12**)
- `make_rescov_for_pair.R` generates two residual covariance matrices for a trait pair:
    - One for the difference between the environmental variances of traits 1 and 2 (**.E1d**)
    - The other for the environmental covariance between them (**.E12**)

Usage examples are available [here](examples.md#genetic-correlation).

### Heritability enrichment
#### Making a SNP info file
A Perl script, `make_snp_info.pl`, is provided to create [the SNP info file](options.md#snp-info-file).

Usage:
- `perl make_snp_info.pl PLINK-bim-file genomic-feature-file output-filename-prefix`
- Both input files should be in plain text format without header lines.
- Genomic-feature-file's first four columns should be chrom, start, end, and category.

Example:
- `perl make_snp_info.pl test.bim test.annot.txt test`
- Input files are [test.bim](https://github.com/jiang18/mph/blob/main/examples/test.bim) and [test.annot.txt](https://github.com/jiang18/mph/blob/main/examples/test.annot.txt)
- An output file named **test.snp_info.csv** will be generated. 

#### Overlapping 