<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="MINQUE for Partitioning Heritability">
    <meta name="author" content="Jicai Jiang">
    
    <link rel="shortcut icon" href="../img/favicon.ico">

    
    <title>Documentation - MPH</title>
    

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/all.css">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/v4-shims.css">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/hack-font@3.3.0/build/web/hack.min.css">
    <link href='//rsms.me/inter/inter.css' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="../css/base.min.css" rel="stylesheet">
    <link href="../css/cinder.min.css" rel="stylesheet">

    
        
        <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/styles/github.min.css">
        
    

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
            <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
        <![endif]-->

    <!-- Google tag (gtag.js) -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-K4MRX3DSPN"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-K4MRX3DSPN');
    </script>

     
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->

            
              <a class="navbar-brand" href="..">MPH</a>
            
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../install/">Install</a>
                    </li>
                
                
                
                    <li class="active">
                        <a href="./">Documentation</a>
                    </li>
                
                
                
                    <li >
                        <a href="../examples/">Examples</a>
                    </li>
                
                
                
                    <li >
                        <a href="../performance/">Performance</a>
                    </li>
                
                
                
                    <li >
                        <a href="../scripts/">Scripts</a>
                    </li>
                
                
                </ul>

            <ul class="nav navbar-nav navbar-right">
                    <li>
                        <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                            <i class="fas fa-search"></i> Search
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../install/">
                            <i class="fas fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="next" href="../examples/">
                            Next <i class="fas fa-arrow-right"></i>
                        </a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#making-a-grm-from-snps">Making a GRM from SNPs</a></li>
            <li class="second-level"><a href="#input">Input</a></li>
                
                <li class="third-level"><a href="#genotype-files">Genotype files</a></li>
                <li class="third-level"><a href="#snp-info-file">SNP info file</a></li>
            <li class="second-level"><a href="#options">Options</a></li>
                
            <li class="second-level"><a href="#output">Output</a></li>
                
            <li class="second-level"><a href="#custom-genotype-coding">Custom genotype coding</a></li>
                
        <li class="first-level "><a href="#making-a-grm-from-grms">Making a GRM from GRMs</a></li>
            <li class="second-level"><a href="#input_1">Input</a></li>
                
            <li class="second-level"><a href="#options_1">Options</a></li>
                
            <li class="second-level"><a href="#output_1">Output</a></li>
                
        <li class="first-level "><a href="#subsetting-a-grm">Subsetting a GRM</a></li>
        <li class="first-level "><a href="#remlminque">REML/MINQUE</a></li>
            <li class="second-level"><a href="#input_2">Input</a></li>
                
                <li class="third-level"><a href="#grm-input">GRM input</a></li>
                <li class="third-level"><a href="#phenotype-file">Phenotype file</a></li>
                <li class="third-level"><a href="#covariate-file">Covariate file</a></li>
            <li class="second-level"><a href="#options_2">Options</a></li>
                
            <li class="second-level"><a href="#output_2">Output</a></li>
                
        <li class="first-level "><a href="#simulation">Simulation</a></li>
        <li class="first-level "><a href="#prediction">Prediction</a></li>
        <li class="first-level "><a href="#general-relationship-matrix">General relationship matrix</a></li>
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Option names may be abbreviated if the abbreviation is unique or is an exact match for some defined option; e.g., <code>--phenotype</code> works the same as <code>--phenotype_file</code>.</p>
</div>
<h2 id="making-a-grm-from-snps">Making a GRM from SNPs</h2>
<h3 id="input">Input</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--binary_genotype_file</code></td>
<td>FILE PREFIX</td>
<td>Required</td>
<td>PLINK bed/bim/fam filename prefix</td>
</tr>
<tr>
<td><code>--snp_info_file</code></td>
<td>FILE</td>
<td>Required</td>
<td>SNP info file in CSV format</td>
</tr>
<tr>
<td><code>--snp_weight_name</code></td>
<td>STRING</td>
<td>Optional</td>
<td>Specify a column name in the SNP info file for weighting SNPs in a GRM. If not set, all variants in the SNP info file will be used with a default weight of 1.</td>
</tr>
<tr>
<td><code>--min_maf</code></td>
<td>FLOAT</td>
<td>Optional</td>
<td>Filter out all variants with minor allele frequency (MAF) less than or equal to the provided threshold [default=0]</td>
</tr>
<tr>
<td><code>--min_hwe_pval</code></td>
<td>FLOAT</td>
<td>Optional</td>
<td>Filter out all variants which have Hardy-Weinberg equilibrium exact test p-value below the provided threshold [default=0]</td>
</tr>
</tbody>
</table>
<h4 id="genotype-files">Genotype files</h4>
<p>MPH uses PLINK bim/fam/bed files for genotypes. The bed/bim/fam file format is described on the <a href="https://www.cog-genomics.org/plink/1.9/formats">PLINK website</a>.</p>
<h4 id="snp-info-file">SNP info file</h4>
<p>The SNP info file is a CSV file with a header line. The first column must list SNP IDs. Each of the other columns corresponds to one GRM, and each row corresponds to a SNP.</p>
<p>If a SNP does not belong to a GRM, leave the corresponding cell <strong>empty</strong>. If a SNP belongs to <strong><em>n</em></strong> GRMs, put <strong>1</strong> in each of the <strong><em>n</em></strong> cells. To partition SNP heritability by <strong>functional annotations</strong>, create a SNP info file in which each column (or GRM) represents a functional annotation category. Functional annotations may overlap; for example, if a SNP belongs to <strong>6</strong> functional annotation categories, put <strong>1</strong> in each of those <strong>6</strong> cells.</p>
<p>In the example of <a href="../examples/#by-chromosomes">partitioning heritability by chromosomes</a>, <a href="https://github.com/jiang18/mph/blob/main/examples/QTL-MAS-2012/chr.snp_info.csv?plain=1"><strong>chr.snp_info.csv</strong></a> has five columns for five chromosomes.</p>
<p><a href="../scripts/#making-a-snp-info-file">A Perl script</a> is provided to create the SNP info file for partitioning heritability across functional annotations.</p>
<h3 id="options">Options</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--make_grm</code></td>
<td>FLAG</td>
<td>Required</td>
<td>To make a GRM from genotypes</td>
</tr>
<tr>
<td><code>--dominance</code></td>
<td>FLAG</td>
<td>Optional</td>
<td>Flag to make a dominance GRM rather than an additive GRM</td>
</tr>
<tr>
<td><code>--num_threads</code></td>
<td>INT</td>
<td>Optional</td>
<td>Number of computational threads to use [default=1]</td>
</tr>
</tbody>
</table>
<h3 id="output">Output</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--output_file</code></td>
<td>FILE PREFIX</td>
<td>Required</td>
<td>Output filename prefix</td>
</tr>
</tbody>
</table>
<p>The GRM calculation generates two files:</p>
<ol>
<li>A file with the suffix <code>.grm.iid</code></li>
<li>A file with the suffix <code>.grm.bin</code></li>
</ol>
<p>For dominance GRMs, the suffixes are modified as follows:</p>
<ol>
<li><code>.d.grm.iid</code></li>
<li><code>.d.grm.bin</code></li>
</ol>
<h3 id="custom-genotype-coding">Custom genotype coding</h3>
<p>MPH includes a feature for computing GRM with custom genotype coding.</p>
<p>To use custom genotype coding:</p>
<ol>
<li>Add 3 columns to the <a href="#snp-info-file">SNP info file</a> specifying the numeric codes for A<sub>1</sub>A<sub>1</sub>, A<sub>1</sub>A<sub>2</sub>, and A<sub>2</sub>A<sub>2</sub> genotypes for each genomic variant.</li>
<li>Name these columns (e.g., header1, header2, header3) in the header line.</li>
<li>When running <code>mph --make_grm</code>, include the option <code>--snp_genotype_coding header1,header2,header3</code> to enable GRM computation using the custom genotype coding specified in the SNP info file.</li>
</ol>
<p>Important notes:</p>
<ul>
<li>The <code>--snp_genotype_coding</code> option requires exactly 3 tokens. Ensure that the first, second, and third token match A<sub>1</sub>A<sub>1</sub>, A<sub>1</sub>A<sub>2</sub>, and A<sub>2</sub>A<sub>2</sub>, respectively.</li>
<li>A<sub>1</sub> and A<sub>2</sub> refer to the alleles in columns 5 and 6 of the <a href="https://www.cog-genomics.org/plink/1.9/formats#bim">PLINK .bim file</a>, respectively.</li>
<li>To provide users with complete control over GRM computation, custom genotype codes are used as provided, without any centering applied.</li>
<li>The <code>--snp_weight_name</code> option remains available for weighting genotype codes.</li>
</ul>
<h2 id="making-a-grm-from-grms">Making a GRM from GRMs</h2>
<h3 id="input_1">Input</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--grm_list</code></td>
<td>FILE</td>
<td>Required</td>
<td>Space-delimited text file listing GRMs (include paths if needed)</td>
</tr>
</tbody>
</table>
<p>The GRM list file is a space-delimited text file, such as <a href="https://github.com/jiang18/mph/blob/main/examples/QTL-MAS-2012/AD.grms.txt"><strong>AD.grms.txt</strong></a>. It has no header and one or two columns. 
The first column must be the file path for each GRM. The second column is optional and can be a short GRM identifier for <code>--make_fore</code> and <code>--make_core</code>. </p>
<h3 id="options_1">Options</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--merge_grms</code></td>
<td>FLAG</td>
<td>Optional</td>
<td>Flag to merge a list of GRMs</td>
</tr>
<tr>
<td><code>--deduct_grms</code></td>
<td>FLAG</td>
<td>Optional</td>
<td>Flag to deduct the subsequent GRMs from the first in a list</td>
</tr>
<tr>
<td><code>--make_fore</code></td>
<td>FLAG</td>
<td>Optional</td>
<td>Flag to make first-order interaction GRMs</td>
</tr>
<tr>
<td><code>--make_core</code></td>
<td>FLAG</td>
<td>Optional</td>
<td>Flag to make GRMs for covariances between random effects</td>
</tr>
<tr>
<td><code>--num_threads</code></td>
<td>INT</td>
<td>Optional</td>
<td>Number of computational threads to use [default=1]</td>
</tr>
</tbody>
</table>
<p>If a SNP is used in multiple GRMs, the SNP will be treated to be multiple identical SNPs in the resulting GRM of <code>--merge_grms</code>.</p>
<h3 id="output_1">Output</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--output_file</code></td>
<td>FILE PREFIX</td>
<td>Required</td>
<td>Output filename prefix</td>
</tr>
</tbody>
</table>
<ul>
<li><code>--merge_grms</code> and <code>--deduct_grms</code> generate one GRM.</li>
<li><code>--make_fore</code> generates <em>n</em>(<em>n</em>+1)/2 GRMs for <em>n</em> GRMs in a GRM list.</li>
<li><code>--make_core</code> generates <em>n</em>(<em>n</em>-1)/2 GRMs for <em>n</em> GRMs in a GRM list.</li>
</ul>
<h2 id="subsetting-a-grm">Subsetting a GRM</h2>
<table>
<thead>
<tr>
<th>Option</th>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--subset_grm</code></td>
<td>FLAG</td>
<td>Required</td>
<td>Flag to initialize the subsetting routine</td>
</tr>
<tr>
<td><code>--binary_grm_file</code></td>
<td>FILE PREFIX</td>
<td>Required</td>
<td>Filename prefix of input GRM to be subsetted</td>
</tr>
<tr>
<td><code>--keep</code></td>
<td>FILE</td>
<td>Required</td>
<td>Text file listing individual IDs to keep in the subset</td>
</tr>
<tr>
<td><code>--output_file</code></td>
<td>FILE PREFIX</td>
<td>Required</td>
<td>Output filename prefix</td>
</tr>
</tbody>
</table>
<p>The file specified by <code>--keep</code> is a text file with one individual ID per line and no header. It determines both which individuals to keep and their order in the output GRM.</p>
<p>The program stops with an error message if any individual IDs listed in the <code>--keep</code> file are not found in the input GRM. The output GRM will have individuals in the exact order specified in the <code>--keep</code> file.</p>
<h2 id="remlminque">REML/MINQUE</h2>
<h3 id="input_2">Input</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--grm_list</code><br><code>--binary_grm_file</code></td>
<td>FILE<br>FILE-PREFIX</td>
<td>Required</td>
<td>Space-delimited text file listing GRMs <strong>and/or</strong> filename prefix of a single GRM (at least one must be specified; include paths if needed)</td>
</tr>
<tr>
<td><code>--phenotype_file</code></td>
<td>FILE</td>
<td>Required</td>
<td>Phenotype file in CSV format</td>
</tr>
<tr>
<td><code>--trait</code></td>
<td>STRING</td>
<td>Required</td>
<td>Single trait name or comma-separated list of trait names that should match the column names in the phenotype file</td>
</tr>
<tr>
<td><code>--error_weight_names</code></td>
<td>STRING</td>
<td>Optional</td>
<td>Single column name or comma-separated list of column names in the phenotype file, specifying individual error variance weights for each corresponding trait</td>
</tr>
<tr>
<td><code>--covariate_file</code></td>
<td>FILE</td>
<td>Optional</td>
<td>Covariate file in CSV format</td>
</tr>
<tr>
<td><code>--covariate_names</code></td>
<td>STRING</td>
<td>Optional</td>
<td>Comma-separated list of covariates to include in the analysis</td>
</tr>
</tbody>
</table>
<h4 id="grm-input">GRM input</h4>
<p>Specify GRMs using one or both of the following options:</p>
<ul>
<li><code>--grm_list</code> with a list file containing one or more GRMs</li>
<li><code>--binary_grm_file</code> to directly specify a single GRM</li>
</ul>
<p>If both options are provided, the GRM specified by <code>--binary_grm_file</code> will be appended to the list from <code>--grm_list</code>.</p>
<p>The GRM list file is a space-delimited text file, such as <a href="https://github.com/jiang18/mph/blob/main/examples/QTL-MAS-2012/chr.grms.txt"><strong>chr.grms.txt</strong></a>. It has no header and one or two columns. 
The first column must be the file path for each GRM. The second column is optional (defaults to 1) and can list an initial variance component (VC) value for each GRM. 
When <code>--binary_grm_file</code> is used with <code>--grm_list</code>, its initial VC value is set to the average of the VC values from the list file.</p>
<h4 id="phenotype-file">Phenotype file</h4>
<p>The phenotype file is a CSV file with a header line. The first column must be the individual ID. The header line needs to contain trait names.</p>
<p>Missing values of phenotypes need to be left empty. <strong>Do not use space, -9, NA, or NaN.</strong></p>
<p>MPH has <code>--error_weight_names</code> to accommodate individual reliabilities (<em>r</em><sup>2</sup>) for pseudo-phenotypes (e.g., de-regressed estimated breeding values). The error weights can be set to 1/<em>r</em><sup>2</sup>-1 and kept as a column of the phenotype file.</p>
<h4 id="covariate-file">Covariate file</h4>
<p>The covariate file is a CSV file with a header line. The first column must be the individual ID. The header line needs to contain covariate names. </p>
<p>Missing values of covariates need to be left empty. <strong>Do not use space, -9, NA, or NaN.</strong></p>
<p>If <code>--covariate_names all</code> is specified, MPH will use as covariates all columns from the 2nd to the last in the covariate file.</p>
<h3 id="options_2">Options</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--minque</code><br><code>--reml</code></td>
<td>FLAG</td>
<td>Required</td>
<td>Flag to run REML or iterative MINQUE. This option can be specified as <code>--minque</code> or <code>--reml</code>, and they are interchangeable.</td>
</tr>
<tr>
<td><code>--save_memory</code></td>
<td>FLAG</td>
<td>Optional</td>
<td>Flag to enable the memory-saving mode</td>
</tr>
<tr>
<td><code>--num_threads</code></td>
<td>INT</td>
<td>Optional</td>
<td>Number of computational threads to use [default=1]</td>
</tr>
<tr>
<td><code>--heritability</code></td>
<td>FLOAT</td>
<td>Optional</td>
<td>Initial heritability value(s) in REML iterations [default=0.5]</td>
</tr>
<tr>
<td><code>--num_grms</code></td>
<td>INT</td>
<td>Optional</td>
<td>Number of the list's first matrices to be used as genomic relationships [default=all]</td>
</tr>
<tr>
<td><code>--num_iterations</code></td>
<td>INT</td>
<td>Optional</td>
<td>Max number of REML iterations [default=20]</td>
</tr>
<tr>
<td><code>--tol</code></td>
<td>FLOAT</td>
<td>Optional</td>
<td>Absolute convergence tolerance for REML log-likelihood [default=0.01]</td>
</tr>
<tr>
<td><code>--num_random_vectors</code></td>
<td>INT</td>
<td>Optional</td>
<td>Number of random probing vectors for stochastic trace estimation [default=100]</td>
</tr>
<tr>
<td><code>--distribution</code></td>
<td>STRING</td>
<td>Optional</td>
<td>Specify whether the stochastic trace estimation uses a <strong>Gaussian</strong> or <strong>Rademacher</strong> distribution [default=Rademacher]</td>
</tr>
<tr>
<td><code>--seed</code></td>
<td>INT</td>
<td>Optional</td>
<td>Random seed in stochastic trace estimation [default=0]</td>
</tr>
</tbody>
</table>
<p>MPH incorporates Fisher's scoring for REML, equivalent to iterative MINQUE, alongside a trust-region dogleg algorithm. <code>--minque</code> mirrors <code>--reml</code>.</p>
<p>To force MINQUE(0) or MINQUE(1), set <code>--num_iterations 1</code>. The second column of the GRM list file should be set to 0 for MINQUE(0) and 1 for MINQUE(1).</p>
<p>The memory-saving mode (<code>--save_memory</code>) is not necessarily slower, especially when file caching is effectively utilized. </p>
<h3 id="output_2">Output</h3>
<table>
<thead>
<tr>
<th>Option</th>
<th>Argument</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--output_file</code></td>
<td>FILE PREFIX</td>
<td>Required</td>
<td>Output filename prefix</td>
</tr>
</tbody>
</table>
<p><code>mph --reml</code> produces four or five output files, as shown below.</p>
<table>
<thead>
<tr>
<th>Filename suffix</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>.mq.blue.csv</td>
<td>Best linear unbiased estimates of fixed effects</td>
</tr>
<tr>
<td>.mq.iter.csv</td>
<td>Summary of REML iterations</td>
</tr>
<tr>
<td>.mq.vc.csv</td>
<td>Variance component estimates</td>
</tr>
<tr>
<td>.mq.py.csv</td>
<td>Residuals of the linear mixed model</td>
</tr>
<tr>
<td>.mq.cor.csv</td>
<td>Estimates and standard errors of correlations between traits. <strong>Exclusively available for multi-trait analyses.</strong></td>
</tr>
</tbody>
</table>
<p><strong>.mq.vc.csv</strong> has the following columns.</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>trait_x</td>
<td>Trait name</td>
</tr>
<tr>
<td>trait_y</td>
<td>Trait name. If the same as trait_x, components correspond to the trait variance; otherwise, the covariance between trait_x and trait_y.</td>
</tr>
<tr>
<td>vc_name</td>
<td>Name of the (co)variance component or GRM</td>
</tr>
<tr>
<td>m</td>
<td>Weighted number of SNPs in the GRM</td>
</tr>
<tr>
<td>var</td>
<td>Estimate of the (co)variance component</td>
</tr>
<tr>
<td>seV</td>
<td>Standard error of the (co)variance component estimate</td>
</tr>
<tr>
<td>pve</td>
<td>Estimate of the proportion of (co)variance explained</td>
</tr>
<tr>
<td>seP</td>
<td>Standard error of the estimate of the proportion of (co)variance explained</td>
</tr>
<tr>
<td>enrichment</td>
<td>Estimate of the per-SNP heritability enrichment</td>
</tr>
<tr>
<td>seE</td>
<td>Standard error of the estimate of the per-SNP heritability enrichment</td>
</tr>
</tbody>
</table>
<p>Additional columns display sampling covariance matrices of estimates for enrichments and for all (co)variance components.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Estimates of PVEs and enrichments are valid only when functional annotation categories do not overlap with one another. If functional categories actually overlap, one more quick computation is needed to <a href="../scripts/#from-vcs-to-enrichments">recompute PVEs and enrichments</a>. </p>
</div>
<h2 id="simulation">Simulation</h2>
<p>Simulating phenotypes based on a list of GRMs</p>
<pre><code class="language-sh">mph --simulate --num_phenotypes 100 --grm_list chr.grms.txt --heritability 0.5 --output pheno
</code></pre>
<p>In the <em>i</em>th row of the GRM list file are GRM(<em>i</em>) and VC(<em>i</em>). MPH simulates total genetic values (<strong>g</strong>) by sampling <strong>g</strong> from N(<strong>0</strong>,<strong>V</strong>) in which <strong>V</strong> is equal to the sum of all GRM(<em>i</em>)*VC(<em>i</em>). MPH further simulates phenotypes by adding an error term (<strong>e</strong>) to <strong>g</strong> based on heritability.</p>
<h2 id="prediction">Prediction</h2>
<p>Empirical best linear unbiased predictions (EBLUPs)</p>
<pre><code class="language-sh">mph --pred --mq_file milk --output milk
</code></pre>
<p>MPH computes EBLUPs using the output of <code>--reml</code> and outputs them to a file with a suffix of <strong>.mq.blup.csv</strong>. For genomic partitioning, EBLUPs are the estimates of direct genomic values. </p>
<p><strong>Multi-trait BLUP is not currently supported.</strong></p>
<h2 id="general-relationship-matrix">General relationship matrix</h2>
<p>The <code>--grm_list</code> file can list any <strong>general</strong> relationship matrix, not necessarily a <strong>genomic</strong> relationship matrix.</p>
<p><a href="../scripts/#grm-inputoutput">Auxiliary scripts</a> are provided to convert a general relationship matrix to the MPH format.</p></div>
        
        
    </div>

    
      <footer class="col-md-12 text-center">
          
          
            <hr>
            <p>
            <small>MPH is licensed under the <a href='https://github.com/jiang18/mph/blob/main/LICENSE'>GPL-3.0 license</a></small><br>
            
            <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</small>
            </p>
          

          
          
      </footer>
    
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="../js/bootstrap-3.0.3.min.js"></script>

    
    <script src="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/highlight.min.js"></script>
        
                <script src="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/languages/r.min.js"></script>
        
    <script>hljs.initHighlightingOnLoad();</script>
    

    <script>var base_url = ".."</script>
    
    <script src="../js/base.js"></script>
    <script src="../search/main.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal">
                    <span aria-hidden="true">&times;</span>
                    <span class="sr-only">Close</span>
                </button>
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
            </div>
            <div class="modal-body">
                <p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
                <form>
                    <div class="form-group">
                        <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>
    </body>

</html>
