<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="MINQUE for Partitioning Heritability">
    <meta name="author" content="Jicai Jiang">
    
    <link rel="shortcut icon" href="../img/favicon.ico">

    
    <title>Examples - MPH</title>
    

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/all.css">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.12.0/css/v4-shims.css">
    <link rel="stylesheet" href="//cdn.jsdelivr.net/npm/hack-font@3.3.0/build/web/hack.min.css">
    <link href='//rsms.me/inter/inter.css' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,700italic,400,300,600,700&subset=latin-ext,latin' rel='stylesheet' type='text/css'>
    <link href="../css/bootstrap-custom.min.css" rel="stylesheet">
    <link href="../css/base.min.css" rel="stylesheet">
    <link href="../css/cinder.min.css" rel="stylesheet">

    
        
        <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/styles/github.min.css">
        
    

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
            <script src="https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js"></script>
            <script src="https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js"></script>
        <![endif]-->

    <!-- Google tag (gtag.js) -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-K4MRX3DSPN"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'G-K4MRX3DSPN');
    </script>

     
</head>

<body>

    <div class="navbar navbar-default navbar-fixed-top" role="navigation">
    <div class="container">

        <!-- Collapsed navigation -->
        <div class="navbar-header">
            <!-- Expander button -->
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            

            <!-- Main title -->

            
              <a class="navbar-brand" href="..">MPH</a>
            
        </div>

        <!-- Expanded navigation -->
        <div class="navbar-collapse collapse">
                <!-- Main navigation -->
                <ul class="nav navbar-nav">
                
                
                    <li >
                        <a href="..">Home</a>
                    </li>
                
                
                
                    <li >
                        <a href="../install/">Install</a>
                    </li>
                
                
                
                    <li >
                        <a href="../options/">Documentation</a>
                    </li>
                
                
                
                    <li class="active">
                        <a href="./">Examples</a>
                    </li>
                
                
                
                    <li >
                        <a href="../performance/">Performance</a>
                    </li>
                
                
                
                    <li >
                        <a href="../util/">Utilities</a>
                    </li>
                
                
                </ul>

            <ul class="nav navbar-nav navbar-right">
                    <li>
                        <a href="#" data-toggle="modal" data-target="#mkdocs_search_modal">
                            <i class="fas fa-search"></i> Search
                        </a>
                    </li>
                    <li >
                        <a rel="prev" href="../options/">
                            <i class="fas fa-arrow-left"></i> Previous
                        </a>
                    </li>
                    <li >
                        <a rel="next" href="../performance/">
                            Next <i class="fas fa-arrow-right"></i>
                        </a>
                    </li>
            </ul>
        </div>
    </div>
</div>

    <div class="container">
        
        
        <div class="col-md-3"><div class="bs-sidebar hidden-print affix well" role="complementary">
    <ul class="nav bs-sidenav">
        <li class="first-level active"><a href="#simulated-datasets">Simulated datasets</a></li>
            <li class="second-level"><a href="#qtl-mas-2012">QTL-MAS 2012</a></li>
                
            <li class="second-level"><a href="#sequence-genotypes">Sequence genotypes</a></li>
                
        <li class="first-level "><a href="#partitioning-heritability">Partitioning heritability</a></li>
            <li class="second-level"><a href="#by-chromosomes">By chromosomes</a></li>
                
            <li class="second-level"><a href="#by-functional-annotations">By functional annotations</a></li>
                
        <li class="first-level "><a href="#dominance-and-epistasis">Dominance and epistasis</a></li>
        <li class="first-level "><a href="#genetic-correlation">Genetic correlation</a></li>
            <li class="second-level"><a href="#genome-wide">Genome-wide</a></li>
                
            <li class="second-level"><a href="#chromosome-wise">Chromosome-wise</a></li>
                
    </ul>
</div></div>
        <div class="col-md-9" role="main">

<div class="admonition note">
<p class="admonition-title">Note</p>
<p>All utility scripts used in the examples are described in <a href="../util/">Utilities</a>.</p>
</div>
<h2 id="simulated-datasets">Simulated datasets</h2>
<h3 id="qtl-mas-2012">QTL-MAS 2012</h3>
<ul>
<li>The data set is available for download <a href="https://github.com/jiang18/mph/raw/main/examples/QTL-MAS-2012.zip">here</a>.</li>
<li><a href="https://bmcproc.biomedcentral.com/articles/10.1186/1753-6561-8-S5-S1">This article</a> describes how the data set was simulated.</li>
<li>3k related animals, 10k SNPs, 5 chromosomes, and 3 traits</li>
</ul>
<h3 id="sequence-genotypes">Sequence genotypes</h3>
<ul>
<li>The data set is available for download <a href="https://doi.org/10.6084/m9.figshare.24432948">here</a>.</li>
<li>Sequence genotypes were simulated by <a href="https://aipl.arsusda.gov/software/genosim/">genosim</a>.</li>
<li>10k unrelated individuals, 5M sequence variants, and 30 chromosomes</li>
<li>Functional annotations were quickly simulated by assigning <a href="https://console.cloud.google.com/storage/browser/_details/broad-alkesgroup-public-requester-pays/LDSCORE/1000G_Phase3_baseline_ldscores.tgz">LDSC baseline annotations</a> to the simulated sequence variants in order.</li>
<li>Phenotypes were simulated using the <a href="https://www.nature.com/articles/ng.3404/figures/4">S-LDSC baseline model enrichment estimates for human traits</a>.<ul>
<li>The enrichment estimates were used to compute variance component (VC) estimates for intercept and 24 main functional annotations.</li>
<li>The VC estimates were set as true values in <a href="../options/#simulation">phenotype simulation</a>.</li>
<li>A small value was added to the intercept's VC to enhance the all-in-one GRM's positive definiteness. </li>
</ul>
</li>
</ul>
<h2 id="partitioning-heritability">Partitioning heritability</h2>
<h3 id="by-chromosomes">By chromosomes</h3>
<p>Partitioning heritability by chromosomes for the <a href="#qtl-mas-2012">QTL-MAS 2012</a> dataset</p>
<ol>
<li>Create a <a href="../options/#snp-info-file">SNP info file</a>: <a href="https://github.com/jiang18/mph/blob/main/examples/QTL-MAS-2012/chr.snp_info.csv?plain=1"><strong>chr.snp_info.csv</strong></a>.</li>
<li><a href="../options/#making-a-grm-from-snps">Make a GRM</a> for each chromosome.</li>
<li>Create a <a href="../options/#grm-list-file">GRM list</a>: <a href="https://github.com/jiang18/mph/blob/main/examples/QTL-MAS-2012/chr.grms.txt"><strong>chr.grms.txt</strong></a>.</li>
<li>Run <a href="../options/#remlminque">REML/MINQUE</a>.</li>
</ol>
<pre><code class="language-sh"># Making GRMs
# Input: geno and chr.snp_info.csv
mkdir chromosomes
for chr in {1..5}
do
mph --make_grm --binary_genotype geno --min_maf 0 --min_hwe_pval 1e-8 --snp_info chr.snp_info.csv --snp_weight $chr --num_threads 10 --out ./chromosomes/$chr
done

# Running REML
# Input: chr.grms.txt, phen.csv, and covar.csv
mph --minque --grm_list chr.grms.txt --phenotype phen.csv --trait milk --error_weight milk_wt --covariate_file covar.csv --covariate_names all --num_threads 10 --out ./chromosomes/milk
</code></pre>
<h3 id="by-functional-annotations">By functional annotations</h3>
<p>Partitioning heritability by functional annotations for the <a href="#sequence-genotypes">sequence genotypes</a> dataset</p>
<pre><code class="language-bash"># 1. Create a new SNP info file including only variants that pass quality control.
# SNPs with an MAF &lt; 0.01 are removed from funct.snp_info.csv, producing qc.funct.snp_info.csv.
snpinfo=&quot;qc.funct.snp_info.csv&quot;

# Read the first line of the SNP info file and split it by a comma to get the annotation list.
IFS=',' read -ra elements &lt; $snpinfo

# 2. Make a GRM for each functional annotation category.
# It may take about one hour to contruct all the GRMs. 
# Note that precomputed GRMs are included in the dataset. Go to the step 3 to save time. 
mkdir grms

# Iterate over the list of annotation categories.
# Use PLINK to extract SNPs in each category for faster I/O.
for ((i = 1; i &lt; ${#elements[@]}; i++)); do
    awk -F ',' -v colname=&quot;${elements[i]}&quot; 'NR == 1 { for (j = 1; j &lt;= NF; j++) if ($j == colname) col = j } NR &gt; 1 &amp;&amp; $col == 1 { print $1 }' $snpinfo &gt;  &quot;${elements[i]}.extract.txt&quot;
    plink --bfile geno --chr-set 30 --extract &quot;${elements[i]}.extract.txt&quot; --threads 14 --make-bed --out ${elements[i]}
    mph --make_grm --binary_geno ${elements[i]} --snp_info $snpinfo --snp_weight ${elements[i]} --num_threads 14 --out ./grms/${elements[i]}
    rm ${elements[i]}.*
done

# 3. Create a GRM list and run REML.
# Create a GRM list.
grmlist=&quot;funct.grms.txt&quot;
if [ -e $grmlist ]; then
    rm $grmlist
fi
for ((i = 1; i &lt; ${#elements[@]}; i++)); do
    echo &quot;grms/${elements[i]} 1&quot; &gt;&gt; $grmlist
done

# Run REML.
mkdir reml
mph --minque --save_memory --grm_list $grmlist --phenotype pheno/hsq0.9.sim.csv --trait 1 --num_threads 14 --out reml/1
</code></pre>
<p>Below is an R script for recomputing the proportions of genetic variance explained and enrichments.</p>
<pre><code class="language-R"># mph_functs.R is documented at https://jiang18.github.io/mph/util/.
source(&quot;mph_functs.R&quot;)
library(data.table)

# Read the SNP info file to get the SNP weighting matrix.
sw = as.matrix(fread(&quot;qc.funct.snp_info.csv&quot;),rownames=1)
sw[is.na(sw)] = 0

# Get the SNP incidence matrix.
# The SNP weighting matrix is often the incidence matrix, but not always.
si = (sw != 0) * 1

# More annotation categories can be included in the incidence matrix.
# For example, a new annotation of gene is derived here.
gene = as.numeric( rowSums(si[, c(&quot;Coding_UCSC&quot;, &quot;Intron_UCSC&quot;, &quot;Promoter_UCSC&quot;, &quot;UTR_3_UCSC&quot;, &quot;UTR_5_UCSC&quot;)]) &gt; 0 )
si = cbind(si, gene)

# Calculate the crossproduct of 'si' and 'sw'.
# The columns of 'sw' should match the rows of 'vcfile'; otherwise, they should be reordered.
index = 1:25; sw = sw[, index]
cp = crossprod(si, sw)

# Number of SNPs in each annotation category of interest
annot_size = colSums(si)
# Total number of SNPs
nsnps = nrow(si)

vcfile = &quot;reml/1.mq.vc.csv&quot;
result = recompute_enrichments(vcfile, cp, nsnps=nsnps, annot.size=annot_size) 
result

# Visualization
library(ggplot2)

result = as.data.frame(result[-1,])
pct = result$prop * 100
pct = round(pct,2)
rownames(result) = paste0(rownames(result),&quot; (&quot;,pct,&quot;%)&quot;)

df = data.frame(FunctionalAnnotation=rownames(result), Enrichment=result$enrichment, SE=result$enrichment.se)
p&lt;- ggplot(df, aes(x=FunctionalAnnotation, y=Enrichment)) + 
  geom_bar(stat=&quot;identity&quot;, color=&quot;black&quot;, 
           position=position_dodge()) +
  geom_errorbar(aes(ymin=Enrichment-SE, ymax=Enrichment+SE), width=.5, position=position_dodge(.9)) 
p&lt;- p + theme(text = element_text(size=22), axis.text.x = element_text(angle = 65, vjust = 1, hjust=1))
p&lt;- p+ geom_hline(yintercept=1, linetype=&quot;dashed&quot;, color = &quot;red&quot;) + xlab(&quot;&quot;) + ylab(&quot;Per-SNP heritability enrichment estimate&quot;) + ggtitle(&quot;Trait 1&quot;)
p
</code></pre>
<p>The following figure will be produced.
<img alt="FunctAnnotTrait1" src="../img/FunctAnnotTrait1.jpg" /></p>
<h2 id="dominance-and-epistasis">Dominance and epistasis</h2>
<p>Decomposing genetic variance into additive, dominance, and epistatic components for the <a href="#qtl-mas-2012">QTL-MAS 2012</a> dataset</p>
<ol>
<li><a href="../options/#making-a-grm-from-snps">Make GRMs from SNPs</a>: one for additive and one for dominance.</li>
<li>Create a <a href="../options/#input_1">GRM list</a> for <code>--make_fore</code>: <a href="https://github.com/jiang18/mph/blob/main/examples/QTL-MAS-2012/AD.grms.txt"><strong>AD.grms.txt</strong></a>.</li>
<li><a href="../options/#options_1">Make first-order interaction GRMs</a>.</li>
<li>Create a <a href="../options/#grm-list-file">GRM list</a> for <code>--minque</code>, listing A, D, AxA, AxD, and DxD: <a href="https://github.com/jiang18/mph/blob/main/examples/QTL-MAS-2012/ADE.grms.txt"><strong>ADE.grms.txt</strong></a>.</li>
<li>Run <a href="../options/#remlminque">REML/MINQUE</a>.</li>
</ol>
<pre><code class="language-sh"># Making one additive GRM and one dominance GRM
# Input: geno and chr.snp_info.csv
mkdir nonadditive

mph --make_grm --binary_genotype geno --min_maf 0 --min_hwe_pval 1e-8 --snp_info chr.snp_info.csv --num_threads 10 --out ./nonadditive/test

mph --make_grm --binary_genotype geno --min_maf 0 --min_hwe_pval 1e-8 --snp_info chr.snp_info.csv --num_threads 10 --out ./nonadditive/test --dom

# Making three first-order interaction GRMs
# Input: AD.grms.txt listing the GRMs that have been made
mph --make_fore --grm_list AD.grms.txt --num_threads 10 --out ./nonadditive/test

# Running REML
# Input: ADE.grms.txt, phen.csv, and covar.csv
mph --minque --grm_list ADE.grms.txt --phenotype phen.csv --trait milk --covariate_file covar.csv --covariate_names all --num_threads 10 --out ./nonadditive/milk
</code></pre>
<h2 id="genetic-correlation">Genetic correlation</h2>
<p>Estimating genetic and environmental correlations for the <a href="#qtl-mas-2012">QTL-MAS 2012</a> dataset</p>
<p>Bivariate REML can be transformed into univariate REML. Though MPH is designed for univariate REML, it can <strong>effectively</strong> do bivariate REML when provided with correct GRMs. </p>
<h3 id="genome-wide">Genome-wide</h3>
<ol>
<li>Make a new phenotype file and a new covariate file for a pair of traits.</li>
<li>Make three GRMs and two residual covariance matrices for a pair of traits.</li>
<li>Run REML using a GRM list file like <a href="https://github.com/jiang18/mph/blob/main/examples/QTL-MAS-2012/bivar.grms.txt"><strong>this</strong></a>. </li>
<li>Compute the genetic correlation estimate from the VC estimates.</li>
</ol>
<pre><code class="language-sh"># The R scripts are documented at https://jiang18.github.io/mph/util/.
mkdir bivarREML

Rscript --no-save make_pheno_for_pair.R phen.csv milk fat ./bivarREML/test

Rscript --no-save make_covariate_for_pair.R covar.csv ./bivarREML/test.covar.csv

mph --make_grm --binary_genotype geno --min_maf 0 --min_hwe_pval 1e-8 --snp_info chr.snp_info.csv --num_threads 10 --out ./bivarREML/test

Rscript --no-save make_grms_for_pair.R ./bivarREML/test ./bivarREML/test

Rscript --no-save make_rescov_for_pair.R ./bivarREML/test ./bivarREML/test

mph --minque --save_mem --grm_list bivar.grms.txt --phenotype ./bivarREML/test.milk.fat.pheno.csv --trait scaled --covariate_file ./bivarREML/test.covar.csv --covariate_names all --num_threads 10 --out ./bivarREML/milk.fat
</code></pre>
<p>Below is an R script for computing between-trait correlation estimates from the <strong>.mq.vc.csv</strong> output file.</p>
<pre><code class="language-R">vc = read.csv(&quot;./bivarREML/milk.fat.mq.vc.csv&quot;)

# Calculate the genetic correlation estimate and SE. 
x1 = vc[1,3]
x2 = vc[2,3]
x12 = vc[3,3]
var_x1 = vc[1,14]
var_x2 = vc[2,15]
var_x12 = vc[3,16]
cov_x1_x2 = vc[1,15]
cov_x1_x12 = vc[1,16]
cov_x2_x12 = vc[2,16]

gcorr = x12 / sqrt(x1 * x2)
var_gcorr = gcorr**2 * ( var_x1/(4 * x1**2) + var_x2/(4 * x2**2) + var_x12/x12**2 + cov_x1_x2/(2*x1*x2) - cov_x1_x12/(x1*x12) - cov_x2_x12/(x2*x12) )
se_gcorr = sqrt(var_gcorr)

print(paste(&quot;Genetic correlation estimate is&quot;, gcorr, &quot;with an SE of&quot;, se_gcorr))

# Calculate the environmental correlation estimate and SE. 
x1 = vc[4,3] + vc[6,3]
x2 = vc[6,3]
x12 = vc[5,3]
var_x1 = vc[4,17] + vc[6,4] **2 + 2*vc[6,17]
var_x2 = vc[6,4] **2
var_x12 = vc[5,18]
cov_x1_x2 = vc[6,17] + vc[6,4] **2
cov_x1_x12 = vc[4,18] + vc[6,18]
cov_x2_x12 = vc[6,18]

ecorr = x12 / sqrt(x1 * x2)
var_ecorr = ecorr**2 * ( var_x1/(4 * x1**2) + var_x2/(4 * x2**2) + var_x12/x12**2 + cov_x1_x2/(2*x1*x2) - cov_x1_x12/(x1*x12) - cov_x2_x12/(x2*x12) )
se_ecorr = sqrt(var_ecorr)

print(paste(&quot;Environmental correlation estimate is&quot;, ecorr, &quot;with an SE of&quot;, se_ecorr))
</code></pre>
<h3 id="chromosome-wise">Chromosome-wise</h3>
<ol>
<li>Make a new phenotype file and a new covariate file for a pair of traits.</li>
<li>Make 15 GRMs (three for each chromosome) and two residual covariance matrices for a pair of traits.</li>
<li>Run REML using a GRM list file like <a href="https://github.com/jiang18/mph/blob/main/examples/QTL-MAS-2012/bivar.chr.grms.txt"><strong>this</strong></a>. </li>
<li>Compute the genetic correlation estimates from the VC estimates.</li>
</ol>
<pre><code class="language-sh">mkdir bivarREML

Rscript --no-save make_pheno_for_pair.R phen.csv milk fat ./bivarREML/test

Rscript --no-save make_covariate_for_pair.R covar.csv ./bivarREML/test.covar.csv

for chr in {1..5}
do
mph --make_grm --binary_genotype geno --min_maf 0 --min_hwe_pval 1e-8 --snp_info chr.snp_info.csv --snp_weight $chr --num_threads 10 --out ./bivarREML/$chr

Rscript --no-save make_grms_for_pair.R ./bivarREML/$chr ./bivarREML/$chr
done

Rscript --no-save make_rescov_for_pair.R ./bivarREML/test ./bivarREML/test

mph --minque --save_mem --grm_list bivar.chr.grms.txt --phenotype ./bivarREML/test.milk.fat.pheno.csv --trait scaled --covariate_file ./bivarREML/test.covar.csv --covariate_names all --num_threads 10 --out ./bivarREML/chr.milk.fat
</code></pre>
<p>Below is an Rscript for computing chromosome-wise genetic correlations between traits.</p>
<pre><code class="language-R">vc = read.csv(&quot;./bivarREML/chr.milk.fat.mq.vc.csv&quot;)

# Calculate the genetic correlation estimate and SE. 
gcorr = matrix(nrow=nrow(vc)/3-1, ncol=2)
colnames(gcorr) = c(&quot;estimate&quot;, &quot;SE&quot;)
rownames(gcorr) = paste0(&quot;Chr&quot;, 1:5)
for (k in 1:nrow(gcorr)) {
    x1 = vc[1+(k-1)*3,3]
    x2 = vc[2+(k-1)*3,3]
    x12 = vc[3+(k-1)*3,3]
    var_x1 = vc[1+(k-1)*3, 8+nrow(vc)+(k-1)*3]
    var_x2 = vc[2+(k-1)*3, 8+nrow(vc)+1+(k-1)*3]
    var_x12 = vc[3+(k-1)*3,8+nrow(vc)+2+(k-1)*3]
    cov_x1_x2 = vc[1+(k-1)*3, 8+nrow(vc)+1+(k-1)*3]
    cov_x1_x12 = vc[1+(k-1)*3, 8+nrow(vc)+2+(k-1)*3]
    cov_x2_x12 = vc[2+(k-1)*3, 8+nrow(vc)+2+(k-1)*3]

    est = x12 / sqrt(x1 * x2)
    var_est = est**2 * ( var_x1/(4 * x1**2) + var_x2/(4 * x2**2) + var_x12/x12**2 + cov_x1_x2/(2*x1*x2) - cov_x1_x12/(x1*x12) - cov_x2_x12/(x2*x12) )
    gcorr[k, 1] = est
    gcorr[k, 2] = sqrt(var_est)
}

# Chromosome-wise genetic correlations
print(gcorr)
</code></pre></div>
        
        
    </div>

    
      <footer class="col-md-12 text-center">
          
          
            <hr>
            <p>
            <small>MPH is licensed under the <a href='https://github.com/jiang18/mph/blob/main/LICENSE'>GPL-3.0 license</a></small><br>
            
            <small>Documentation built with <a href="http://www.mkdocs.org/">MkDocs</a>.</small>
            </p>
          

          
          
      </footer>
    
    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
    <script src="../js/bootstrap-3.0.3.min.js"></script>

    
    <script src="//cdn.jsdelivr.net/gh/highlightjs/cdn-release@9.18.0/build/highlight.min.js"></script>
        
    <script>hljs.initHighlightingOnLoad();</script>
    

    <script>var base_url = ".."</script>
    
    <script src="../js/base.js"></script>
    <script src="../search/main.js"></script>

    <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal">
                    <span aria-hidden="true">&times;</span>
                    <span class="sr-only">Close</span>
                </button>
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
            </div>
            <div class="modal-body">
                <p>
                    From here you can search these documents. Enter
                    your search terms below.
                </p>
                <form>
                    <div class="form-group">
                        <input type="text" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>
    </body>

</html>
